### DISCUSSION TITLE
Error 500 after reload cache

### DISCUSSION THREAD

**Discussion by VOOM108** [kimai/kimai#2634] [Questions & Answers]
I have been using kimai2 1.1 for a long time without updating, since most larger changes tended to break everything for me. Now I needed to change invoices the first time after I have updated Linux from 14.04 to 20.04 (which is the only change I can think of) and after clearing the kimai2 cache I got a 500 error on any page. 

I was able to restore last night's backup, but how can I prevent that from happening in the future, as I need to change invoices or do other things, that require clearing the cache? 

Also: I did a fresh install and wanted to start using that, letting 1.1 sit there for the data to still be accessible. Will newer versions be more robust when it comes to little unavoidable things possibly breaking the installation big time?

**Comment by kevinpapst** [comment]
I am literally managing hundreds of installations (not only in the cloud) and they didn't break since the first version. Kimai is used thousands of times all over the world and updates (seem to) work if you stick to the documented way.

Did you consider that this is PEBKAC 😁 ? Totally no offense! Everything you need is documented => but hosting is simply a complicated topic. If things break and you can't resolve them, then it might that you are missing knowledge.

You can use the Cloud version if you are not sure about hosting issues.

**Reply by VOOM108** [reply]
The only way I am going to learn relevant aspects is by asking and learning. I do read every documentation page first, usually several times, and I follow the steps by the letter. If something still goes wrong, I come here and ask. 

I have even with my little knowledge been able to manage quite a few applications in hosted as well as virtual server or cloud server environments. My main line of work are LMSs and I manage Moodle and ILIAS nicely on my cloudserver and many other places whereever customers want them. The list of applications I have worked with so far is long and I have been able to solve most problems with documentations and of course the community. I have never experienced an application that breaks as easily as kimai does. That may still be PEBKAC-related, as I have my troubles with the cli and git. 

Now in this case there was not much to find in the documentation, as all I did was what I have done many times before: clear the kimai cache with the (v1.1) two cli lines. The "only" thing that has changed in the meantime is the Linux env. Of course that is no small change, esp. with the version jump, but there is nothing in the docs that explains why the clearing of the cache now would break the installation, thus: my question what could be happening there. 

And again: all the other things installed in my cloudserver are up and running for months after the server update. Little things to fix but nothing I could not handle....

**Reply by kevinpapst** [reply]
1.1 was a very early version of Kimai , 2 years old now. 

> I have never experienced an application that breaks as easily as kimai does. 

Well, I agree that you have to follow the update steps thoroughly, but then it simply works.

And on the other hand: I don't care - use another software then. Coming here to get my attention and ask for my personal lifetime to help you for free and then ramble how broken Kimai is won't get you any sympathy.

If you want help then follow these simple rules:
Share your problem. Share what you already tried to fix the issue. Post relevant content from the log file. Be nice.
Posting "it doesn't work" is only annoying for both of us.

And if Kimai doesn't work after clearing the cache without doing anything else it is in 99% of all cases caused by wrong file permissions. This is clearly documented and can easily be spotted in the log file.

**Reply by VOOM108** [reply]
Sorry if I appeared to mean it that way. I do appreciate kimai and have been using it since early kima v1 versions. And I don't say that it is broken, but that it breaks easily compared to other apps I use. And I admitted that I have trouble with the cli and git, but still I usually get there eventually. 

I think it is also not so nice, if there is a place for commmunity-discussions that states "help by the community" to get your understandable annoyance with "dumb questions" or a "RTFM" as a first reaction. I was thinking that other users surely can help as well with maybe more regular stuff. I need a nudge sometimes, to remember things I already knew or did before, but have not looked at for months or even years. Heck, I am not 30 anymore ;)  

Sure, I will try to find the log and check there, but many times logfiles are TMI for me, in order to "easily spot" things. What seems easy to the developer who knows his app by heart, things may seem easy or "well I have done that many times and it is documented". Even the documentation is - while it is great that it is - by now so thorough, that as a user one does not always where to look for specific information.

**Reply by kevinpapst** [reply]
> Heck, I am not 30 anymore ;)

Me neither, so I know that problem hahaha

Check the helper script at the bottom of: https://www.kimai.org/documentation/cache.html
This helps massively to reduce the cognitive load when trying to remember how the cache and file permissions work 😁 I use it myself in all custom installations.

Logfile is in var/logs/prod.log - if that doesn't exist or isn't updated then your file permissions might be the problem.
Without logs help is almost impossible (besides that > 80% of all "500 errors" are caused by wrong file permissions).

**Reply by VOOM108** [reply]
I think I have trouble understanding some aspects specifically concerning permissions under Plesk. In the kimai2 installation documentation for Plesk it says: 

`chown -R psacln:psaserv kimai2`

Now I am confused by psacln in Plesk being the group that is readable by opening files in the browser and psaserv is the group for server-side files. Usually, setting the account-username as owner and adjusting the group to either psacln or psaserv works. Now I am not sure what to do with the line above and I might have set up permissions the wrong way. 

(BTW I have started to keep 1.1 purely as an archive and set up the fresh install anew with a clearer concept, so no need to fix that anymore) 

For the fresh install I used the documentation for Plesk by the letter, but still I think some permissions may be wrong which could produce trouble when updating, installing invoices and plugins and clearing the cache afterwards. To fix the permissions I probably mostly need to understand the above line and what to put into owner:group specifically. Once I get that right I should be able to go on on my own.

**Reply by kevinpapst** [reply]
The "only" problem with the permissions is: Kimai creates cache files during the clear cache script, but also during runtime (eg. for caching database results).
Now whe webserver/php uses the group (here psaserv) for creating new files (cache/invoices ...). But when you call the "clear cache" script manually with your own user, the cache files and directories are created with a different (wrong) owner. Therefor you have to set the permissions after clearing the cache, so PHP can write as well.
Either you allow anyone to create files in these directories (bad solution) or you set the correct owner:group combination and allow the PHP process to write to these directories.

Long story short:

I have never used Plesk, so I am not able to help there unfortunately 😢 

As you say: you need to find the correct permissions once and then add them to the clear cache script and you can forget about it. So maybe there is a general Plesk "file permissions and cache" documentation?

**Reply by VOOM108** [reply]
Ok, now I got it. Luckily I seem to have set the permissions right during the installation because the psacln:psaserv did not make sense to me. The same, being "andreas:psacln" I now used for fixing the permissions after clearing the cache. (:psacln was not sufficent because I was using a root account just now). What helped was understanding what happens during clearing - and looking out for a before/after in the var folders. :)

**Reply by kevinpapst** [reply]
Awesome! Now coming back to the PEBKAC 😈 no just kidding, if you want to provide some additional infos for the Plesk-Kimai docs, maybe we can save other ones from the same problems?

**Reply by VOOM108** [reply]
This was rather a result of trial and error, but in the least I suppose the "psacln:psaserv" may be wrong. mybe sth like "<yourwebuser>:psacln" because I am not sure, whether psaserv is the right choice for all of the folders. And maybe add that the same permissions need to be used when fixing the permissions later on, even though that may seem redundant. Whoever wrote that piece of documentation may be better suited to correct or supplement it with better knowledge of cli under plesk.
